alibi/explainers/shap.py

```diff
@@ -0,0 +1,476 @@
+            return DenseData(
+                background_data.values.reshape(1, len(background_data)),
+                list(background_data.index),
+                groups,
+            )
+
+        return background_data
+
+    @_get_data.register(scipy.sparse.spmatrix)
+    def _(self, background_data, *args) -> Union[shap.common.Data, scipy.sparse.spmatrix]:
+        """
+        Initialises background data if user passes a sparse matrix as input. If the
+        user specifies feature grouping, then the sparse array is converted to a dense
+        array. Otherwise, the original array is returned and handled internally by shap
+        library.
+        """
+
+        if self.use_groups:
+            logging.warning(
+                "Grouping is not currently compatible with sparse matrix inputs. "
+                "Converting background data sparse array to dense matrix."
+            )
+            background_data = background_data.toarray()
+            return DenseData(
+                background_data,
+                *args,
+            )
+
+        return background_data
+
+    @_get_data.register(pd.core.frame.DataFrame)
+    def _(self, background_data, *args, **kwargs) -> Union[shap.common.Data, pd.core.frame.DataFrame]:
+        """
+        Initialises background data if user passes a pandas.core.frame.DataFrames as input.
+        If the user has specified groups and given a data frame, initialise shap.common.DenseData
+        explicitly as this is not handled by shap library internally. Otherwise, data initialisation,
+        is left to the shap library.
+        """
+
+        _, groups, weights = args
+        if self.use_groups:
+            logging.info("Group names are specified by column headers, group_names will be ignored!")
+            keep_index = kwargs.get("keep_index", False)
+            if keep_index:
+                return DenseDataWithIndex(
+                    background_data.values,
+                    list(background_data.columns),
+                    background_data.index.values,
+                    background_data.index.name,
+                    groups,
+                    weights,
+                )
+            else:
+                return DenseData(
+                    background_data.values,
+                    list(background_data.columns),
+                    groups,
+                    weights,
+                )
+        else:
+            return background_data
+
+    def explain(self,
+                X: Union[np.ndarray, pd.DataFrame, pd.Series, scipy.sparse.spmatrix],
+                explanation_type: str = 'global',
+                cat_vars_start_idx: Sequence[int] = None,
+                cat_vars_enc_dim: Sequence[int] = None,
+                **kwargs) -> Explanation:
+        """
+        Explains the instances in the array X.
+
+        Parameters
+        ----------
+        X
+            Array with instances to be explained.
+        explanation_type
```

Comment:
`explanation_type` shouldn't be a user argument, this is specified as an atribute of the explainer to signal what explanation types are available.